### PR TITLE
add firebase_element demo

### PR DESCRIPTION
**Pull request by zoechi**

### CODE REVIEW

**Comment by zoechi** [comment]
Should there be any copyright header added to the files?
Most files are auto-generated from my script anyway so this would'nt cause much work.

**Comment by jakemac53** [comment]
Yes, please add copyright notices if you don't mind. You can copy the ones from the main repo.

File: demo/.gitignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+.github_token
```
Comment (jakemac53) [inline demo/.gitignore:1]: probably want to add:

.packages
packages
build
.pub

File: demo/web/firebase_element/app_element.html [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+    <div class="horizontal center-justified layout">
+      <div class="vertical-section">
+        <h3>Dinosaurs by Height</h3>
+        <template is="dom-bind">
+          <firebase-collection
+              order-by-child="height"
+              location="https://dinosaur-facts.firebaseio.com/dinosaurs"
+              data="{{dinosaurs}}"></firebase-collection>
+          <template is="dom-repeat" items="[[dinosaurs]]" as="dinosaur">
```
Comment (jakemac53) [inline demo/web/firebase_element/app_element.html]: Just curious, is this working properly for  you? I was porting the firebase-collection test and had some issues with a dom-repeat inside a dom-bind.

File: demo/web/firebase_element/x_login.dart [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  bool computePasswordHidden(String provider) => provider != 'password';
+
+  @eventHandler
+  bool computeCreateUserDisabled(String email, String password) =>
+      email == null || email.isEmpty || password == null || password.isEmpty;
+
+  @eventHandler
+  bool computeChangePasswordDisabled(
+          String email, String password, String newPassword) =>
+      email == null ||
```
Comment (jakemac53) [inline demo/web/firebase_element/x_login.dart:88]: nit: format, or if this is what the formatter did then file a bug :)

File: demo/web/firebase_element/x_login.dart [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+          password.isEmpty ||
+          newPassword == null ||
+          newPassword.isEmpty;
+
+  @eventHandler
+  bool computeResetPasswordDisabled(String email, String password) =>
+      email == null || email.isEmpty || password == null || password.isEmpty;
+
+  @eventHandler
+  bool computeRemoveUserDisabled(email, password) =>
```
Comment (jakemac53) [inline demo/web/firebase_element/x_login.dart]: Might as well put the type annotations here as well

File: demo/web/firebase_element/x_login.dart [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+  }
+
+  FirebaseAuth get _firebaseLogin => $['firebaseLogin'];
+
+  @eventHandler
+  void logout([_, __]) => _firebaseLogin.logout();
+
+  @eventHandler
+  errorHandler(dom.Event e, var detail) =>
+      message = 'Error: ${detail['message']}';
```
Comment (jakemac53) [inline demo/web/firebase_element/x_login.dart]: This should be `set(message, 'Error: ${detail['message']}')`, or call notifyPath. You could also define a getter/setter for the property (and call notifyPath in the setter). Ditto in `userSuccessHandler`.

File: demo/web/firebase_element/x_pretty_json.dart [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+  XPrettyJson.created() : super.created();
+  @property Map object;
+
+  @eventHandler
+  String prettify(Map object) {
+    if (object == null) {
+      return '';
+    }
+
+    return new JsonEncoder.withIndent('  ').convert(object);
```
Comment (jakemac53) [inline demo/web/firebase_element/x_pretty_json.dart:28]: I did not know about `withIndent`, nice!

**Comment by jakemac53** [comment]
looks great, just some tiny nits and the copyright notices

**Comment by jakemac53** [comment]
> Most files are auto-generated from my script anyway so this would'nt cause much work.

Curious to know more about this script ;)

File: demo/web/firebase_element/x_login.dart (hunk 4 above)
Comment (zoechi) [inline demo/web/firebase_element/x_login.dart]: I created a bug report dart-lang/dart_style#434

File: demo/web/firebase_element/app_element.html (hunk 2 above)
Comment (zoechi) [inline demo/web/firebase_element/app_element.html]: Seems to work, a list of dinosaurs is shown but I created `<app-element>` so I can avoid `<dom-bind>` but I missed this one.

The intention for `<app-element>` also was to be able to additionally integrate all demos in one single application page with a menu to switch between all the demos. 
Polymer.js uses `<iframe>` for this but I don't think this is the optimal solution for Dart.
I just run into an issue with this approach because https://github.com/dart-lang/polymer_elements/blob/master/lib/src/paper-styles/demo-pages.html doesn't work this way. It looks like it needs to be changed from `<style is="custom-style">` to `<dom-element id="some-id">` 
What do you think about how to proceed?

Comment (jakemac53) [inline demo/web/firebase_element/app_element.html]: https://github.com/PolymerElements/paper-styles/blob/master/demo/index.html is an example of how they are using the `demo-pages.html`

**Comment by zoechi** [comment]
> Curious to know more about this script ;)

I plan to commit it to the repo but the code is still too ugly and mostly experimental state. I'll port a few more examples from the generated stubs to see if the result is satisfying and then I'll clean up the code and create a PR.
I also included check for changes but couldn't test it, because there where none yet.

File: demo/web/firebase_element/app_element.html (hunk 2 above)
Comment (zoechi) [inline demo/web/firebase_element/app_element.html]: Yup, but some styles are not applied when I wrap the demo inside an `<app-element>`.

To import styles in the entry page you use `<style is="custom-style"...>` in the imported file, to import the styles in an element the styles need to be wrapped in `<dom-element id="demo-pages"><template>`.
See also: https://www.polymer-project.org/1.0/docs/devguide/styling.html#style-modules
Some styles in `demo-pages.html` as it currently is, are only applied on the entry page and only target elements in the entry page, but not elements inside other elements.
I modified `polymer_elements/lib/src/paper-styles/demo-pages.html` locally from `<style is="custom-style">` to `<dom-module id="demo-pages><template>` and the styles are applied inside `<app-element>`.

Creating a copy of the about 30 lines of styles in `demo-pages.html` file wrapped as `<dom-module id="demo-pages">` (inside `demo-pages.html` or in a new file) would fix this.
What do you think?

Comment (jakemac53) [inline demo/web/firebase_element/app_element.html]: So if we modify `polymer_elements/lib/src/paper-styles/demo-pages.html` it will just get bashed over the next time we update, which I would like to avoid. You could for now just have a custom `demo-pages.html` file in the `demo` project though. Really this change needs to be made upstream, but as it would presumably break a number of JS demos that might be an issue.

**Comment by zoechi** [comment]
The latest update should address all your comments.
I added a `lib/styles/demo_elements.*` element containing the styles from `demo-pages.html`. We could leave it this way or maybe you have a better idea how to deal with this.

File: demo/web/firebase_element/app_element.html (hunk 2 above)
Comment (zoechi) [inline demo/web/firebase_element/app_element.html]: Mit freundlichen Grüßen

Günter Zöchbauer
guenter@gzoechbauer.com
+43 (699) 10 18 87 15

**Comment by jakemac53** [comment]
I left some messages on the newest commit, I always forget those don't show up in the PR :/

edit: I guess they do show up :)

**Comment by zoechi** [comment]
Thanks for the hint with `notify`. Wasn't quite sure how and when to use it. I tried it because the example didn't work as expected, but in the end turned out to be another mistake, just forgot to remove it later.

> I left some messages on the newest commit, I always forget those don't show up in the PR :/

I'm not quite sure how to handle PRs and successive updates. 
For later elements there probably won't as much communication anyway.
I get email notifications at least for your comments on commits (just takes some time until they arrive).

**Comment by zoechi** [comment]
> missed this before, but this attribute is missing the quotes around [[dinosaursScores]]. 

Good catch!

**Comment by jakemac53** [comment]
@zoechi looks like I forgot to ever merge this - let me know when things are ready to be looked at again (looks like you have just continued adding more demos which is fine)

**Comment by zoechi** [comment]
Its fine, I need to revisit them with the released Polymer version. I hoped for an elements update before that. I saw you updated apigen, which makes it even more likely to get updated elements soon.

**Comment by zoechi** [comment]
@jakemac53 Currently I'm done. Updating the elements would be great, then I'll revisite the demos where I encountered issues.

File: demo/pubspec.yaml [hunk 7]
```diff
@@ -0,0 +1,264 @@
+name: polymer_elements_demo
+version: 0.1.0
+author: Polymer.dart Authors <web-ui-dev@dartlang.org>
+description: Polymer polymer-elements for Dart examples.
+homepage: https://www.dartlang.org/polymer/
+dependencies:
+  browser: any
+  polymer: 1.0.0-rc.2
```
Comment (jakemac53) [inline demo/pubspec.yaml]: ^1.0.0-rc.2

File: demo/tool/README.md [hunk 8]
```diff
@@ -0,0 +1,215 @@
+# Check for updated demo files in GitHub organizations and repositories
+
+## Update tool
+
+### </a>Prepare
```
Comment (jakemac53) [inline demo/tool/README.md:5]: Shouldn't need any explicit links (ditto below). I don't think these work in github markdown?

File: demo/tool/update.dart [hunk 9]
```diff
@@ -0,0 +1,791 @@
...
+    await Future.wait(tasks);
+  }
+}
+
+/// Creates a file name `some_element_demo` from a tag name `some-element-demo`
+String _tagNameToFileName(String tagName) => tagName.replaceAll('-', '_');
+
+/// Creates a library name `web.some_element.some_element_demo` from a path
+/// `web/some_element/some_element_demo`
+String _pathToLibraryName(String path) => '${path.split('/').join('.')}.';
```
Comment (jakemac53) [inline demo/tool/update.dart]: extra `.` at the end here it looks like

File: demo/tool/update_status.dart [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+      repositories = repositories.where(
+          (repo) => repo['change'] == 'RepositoryChange.${repositoryFilter}');
+    } else if (repositoryFilter == 'any-changes') {
+      repositories = repositories.where((repo) =>
+          repo['change'] != 'RepositoryChange.none' ||
+              (repo['files'] as List)
+                  .where((file) => file['change'] != 'FileChange.none')
+                  .isNotEmpty);
+    }
+    return repositories.map((repo) => convertToJs(repo)).toList();
```
Comment (jakemac53) [inline demo/tool/update_status.dart:114]: you shouldn't have to manually use `convertToJs` manually, if you do then its probably a bug

File: demo/tool/update_status.dart [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+  }
+
+  @reflectable
+  Object getRepositoryData(Map data, String organization, Object repository) {
+    return convertToJs(_getRepositoryData(data, organization, repository));
```
Comment (jakemac53) [inline demo/tool/update_status.dart:119]: ditto

File: demo/web/all_demos.dart [hunk 12]
```diff
@@ -0,0 +1,404 @@
...
+//  import 'neon_animation/dropdown/index.dart';
+//  import 'neon_animation/grid/index.dart';
+//  import 'neon_animation/list/index.dart';
+//  import 'neon_animation/load/index.dart';
+//  import 'neon_animation/reprojection/index.dart';
+//  import 'neon_animation/tiles/index.dart';
+
+
+/// Silence analyzer [DemoElements], [PaperItem], [PaperDrawerPanel],
+/// [PaperHeaderPanel], [IronList], [PaperToolbar],
```
Comment (jakemac53) [inline demo/web/all_demos.dart:134]: probably makes sense to just condense all of these onto as few lines as possible

File: demo/web/all_demos.dart [hunk 13]
```diff
@@ -0,0 +1,404 @@
...
+/// [PaperToastDemo]
+/// [PaperToggleButtonDemo]
+/// [PaperToolbarDemo]
+/// [PaperTooltipDemo]
+@PolymerRegister('all-demos')
+class AllDemos extends PolymerElement {
+  AllDemos.created() : super.created();
+  List<DemoElementItem> __demos;
+  @property List<DemoElementItem> get demos => __demos;
+  set demos(value) {}
```
Comment (jakemac53) [inline demo/web/all_demos.dart]: remove?

File: demo/web/all_demos.dart [hunk 14]
```diff
@@ -0,0 +1,404 @@
...
+/// [PaperStylesDemo]
+/// [PaperTabsDemo]
+/// [PaperToastDemo]
+/// [PaperToggleButtonDemo]
+/// [PaperToolbarDemo]
+/// [PaperTooltipDemo]
+@PolymerRegister('all-demos')
+class AllDemos extends PolymerElement {
+  AllDemos.created() : super.created();
+  List<DemoElementItem> __demos;
```
Comment (jakemac53) [inline demo/web/all_demos.dart]: any particular reason for the double underscore?

File: demo/web/all_demos.dart [hunk 15]
```diff
@@ -0,0 +1,404 @@
...
+    notifyPath('demos', __demos);
+    _demoList.selectItem(convertToJs(demos.first));
+    _loadDemo(demos.first);
+  }
+
+  IronList get _demoList => $['demolist'] as IronList;
+
+  @reflectable
+  void demoClickHandler(dom.Event event, [_]) {
+    final item = convertToDart(_demoList.jsElement.callMethod(
```
Comment (jakemac53) [inline demo/web/all_demos.dart:260]: Mind referencing https://github.com/dart-lang/polymer_elements/issues/90 in a comment above this?

File: demo/web/firebase_element/index.html [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+    <meta name="viewport" content="width=device-width, minimum-scale=1.0, initial-scale=1, user-scalable=yes">
+
+    <title>firebase-element-demo</title>
+
+    <script src="packages/web_components/webcomponents-lite.min.js"></script>
+  </head>
+  <body unresolved>
+    <firebase-element-demo></firebase-element-demo>
+
+    <script src="packages/browser/dart.js"></script>
```
Comment (jakemac53) [inline demo/web/firebase_element/index.html]: I think this has to come after the dart script tag

File: demo/web/firebase_element/x_login.dart [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+    }
+
+    if (provider == 'password') {
+      parameters = parameters ?? {};
+      parameters['email'] = email;
+      parameters['password'] = password;
+    }
+
+    //_firebaseLogin.login(parameters, null);
+    _firebaseLogin.jsElement.callMethod('login', [parameters, null]);
```
Comment (jakemac53) [inline demo/web/firebase_element/x_login.dart:51]: please mention https://github.com/dart-lang/polymer_elements/issues/59 here

File: demo/web/firebase_element/x_login.dart [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+  }
+
+  FirebaseAuth get _firebaseLogin => $['firebaseLogin'];
+
+  @reflectable
+  void logout([_, __]) => _firebaseLogin.logout();
+
+  @reflectable
+  errorHandler(dom.Event e, var detail) =>
```
Comment (jakemac53) [inline demo/web/firebase_element/x_login.dart]: no need for `var`

File: demo/web/firebase_element/x_login.dart [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+  bool computePasswordHidden(String provider) => provider != 'password';
+
+  @reflectable
+  bool computeCreateUserDisabled(String email, String password) =>
+      email == null || email.isEmpty || password == null || password.isEmpty;
+
+  @reflectable
+  bool computeChangePasswordDisabled(
+          String email, String password, String newPassword) =>
+      email == null ||
```
Comment (jakemac53) [inline demo/web/firebase_element/x_login.dart:88]: formatter bug?

**Comment by jakemac53** [comment]
@zoechi have you used rietveld before? For a cl of this size it would help me a lot when reviewing (and its also easier to see changes/responses over time). It is also to big for the github UI to handle lol.

File: demo/web/google_castable_video/google_castable_video_demo.dart [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+  void timeUpdate(dom.CustomEvent event, [_]) {
+    var duration = _video.duration;
+    var currentTime = event.detail['currentTime'];
+    set('progress', '${currentTime * (100 / duration)}');
+  }
+
+  //listen for casting event to change icon
+  @reflectable
+  casting(dom.CustomEvent event, [_]) {
+    ($['cast'] as dom.ButtonElement).innerHtml =
```
Comment (jakemac53) [inline demo/web/google_castable_video/google_castable_video_demo.dart]: Seems like this should just be a property and a binding? If it matches the JS demo I guess that is ok though.

File: demo/web/google_drive/google_drive_demo.dart [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+class GoogleDriveDemo extends PolymerElement {
+  GoogleDriveDemo.created() : super.created();
+
+  @property String statusMessage;
+  @property List queueUploadList;
+  @property List uploadedList;
+
+  @reflectable
+  void uploadStarted(dom.CustomEvent event, [_]) {
+    set('queueUploadList', (event.detail.files as List).toList());
```
Comment (jakemac53) [inline demo/web/google_drive/google_drive_demo.dart:31]: isn't `event.detail` a `JsObject` or a `Map` (ditto below).

File: demo/tool/README.md (hunk 8 above)
Comment (zoechi) [inline demo/tool/README.md:5]: I made this to be able to link to the caption from farther below. I looked it up somewhere and don't know about another way. I tried it by opening the file using the "View" button and it works.

File: demo/tool/update_status.dart (hunk 10 above)
Comment (zoechi) [inline demo/tool/update_status.dart:114]: It fails at line 80 if I remove it

File: demo/web/all_demos.dart (hunk 12 above)
Comment (zoechi) [inline demo/web/all_demos.dart:134]: Seems like a lot of work especially when one gets added later somewhere in the middle and all the linebreaks below have to be adjusted.

File: demo/web/all_demos.dart (hunk 13 above)
Comment (zoechi) [inline demo/web/all_demos.dart]: Oh, I was fighting with some bug at some point. This is unintended leftover.

File: demo/web/all_demos.dart (hunk 14 above)
Comment (zoechi) [inline demo/web/all_demos.dart]: This was a backing field for a getter/setter with a private setter, therefore the name `_emos` was already taken. Can't remember exactly why I did this. Just a leftover I forgot to remove.

File: demo/web/firebase_element/index.html (hunk 16 above)
Comment (zoechi) [inline demo/web/firebase_element/index.html]: Right, good catch :+1:

File: demo/web/firebase_element/x_login.dart (hunk 18 above)
Comment (zoechi) [inline demo/web/firebase_element/x_login.dart]: variables, parameters and functions without type annotations always look like broken/experimental code to me. Removed anyway :)

File: demo/web/firebase_element/x_login.dart (hunk 19 above)
Comment (zoechi) [inline demo/web/firebase_element/x_login.dart:88]: I think this is correct / as intended. 
After `=>` the indentation is reduced to be relative to the start of the parent expression and after `email == null ||` indentation is again increased to indicate that it continues the previous line.
Actually, I don't feel entitled to discuss the results of the formatter. I'm just glad I don't have to make all these decisions myself :sweat:

File: demo/web/google_castable_video/google_castable_video_demo.dart (hunk 20 above)
Comment (zoechi) [inline demo/web/google_castable_video/google_castable_video_demo.dart]: This is actually a response to 

> @zoechi have you used rietveld before? For a cl of this size it would help me a lot when reviewing (and its also easier to see changes/responses over time). It is also to big for the github UI to handle lol.

because this comment of yours has no reply button.
I haven't used Rietveld before but if you can point out where to start I'd give it a try.

Comment (zoechi) [inline demo/web/google_castable_video/google_castable_video_demo.dart]: > Seems like this should just be a property and a binding? If it matches the JS demo I guess that is ok though.

Right. I usually changed it when I encountered something like that in the JS demos. This was an oversight.

The JS demos are often not quite polymery. Sometimes they don't even use `dom-bind`.
Moving them into components "forced" me to improve them. IMHO the demos should not only demonstrate how the components look like but also how they are being used in a Polymer application. Not every Polymer element is used on the entry page :-/ </ranting>

File: demo/web/google_drive/google_drive_demo.dart (hunk 21 above)
Comment (zoechi) [inline demo/web/google_drive/google_drive_demo.dart:31]: There is no `google-drive` component yet in dart-lang/polymer_elements. This code is not tested.

File: demo/web/firebase_element/x_login.dart (hunk 19 above)
Comment (jakemac53) [inline demo/web/firebase_element/x_login.dart:88]: Haha fair enough, I certainly am not suggesting that this code should be changed from what the formatter does. At the most I would just file a bug with the formatter if we thought that this was "wrong".

File: demo/web/google_drive/google_drive_demo.dart (hunk 21 above)
Comment (jakemac53) [inline demo/web/google_drive/google_drive_demo.dart:31]: ah ok that would explain it ;)

File: demo/web/iron_a11y_keys/x_key_aware.dart [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+import 'package:polymer_elements/iron_a11y_keys.dart';
+
+/// Silence analyzer [IronA11yKeys]
+@PolymerRegister('x-key-aware')
+class XKeyAware extends PolymerElement {
+  XKeyAware.created() : super.created();
+
+  String __pressed = '';
+  @property String get pressed => __pressed;
+  set _pressed(String value) {
```
Comment (jakemac53) [inline demo/web/iron_a11y_keys/x_key_aware.dart:25]: any particular reason to make the setter private? (ditto for some other elements below)

Comment (zoechi) [inline demo/web/iron_a11y_keys/x_key_aware.dart:25]: In the JS example the property is read-only and I just imitated it

``` js
 properties: {
      pressed: {
        type: String,
        readOnly: true,
        value: ''
      },
```

**Comment by jakemac53** [comment]
Sorry for all the flip-flopping, but the more that I look at this the more I am thinking that we should actually go back to the old way of having a separate repo for the demos.

My primary concern is maintenance of the demos. This is a lot of code which is going to have a high churn rate, and making it be part of the primary package adds a major barrier to updating the elements in a timely fashion. If these live in a separate repo then they can be updated on a different timeline without slowing down the overall updating process.

File: demo/web/iron_a11y_keys/x_key_aware.dart (hunk 22 above)
Comment (jakemac53) [inline demo/web/iron_a11y_keys/x_key_aware.dart:25]: ah ok sure that makes sense then

**Comment by zoechi** [comment]
No prob. It's an experiment. I'm still trying to figure what strategy with examples works best in what situation (for example also in bwu_datagrid). Please create a repo for the demos and I create a new PR.

**Comment by jakemac53** [comment]
@zoechi I think it actually makes sense to have this as a repo under your personal account, and we will link to it from the readme? That way you can maintain credit for it and also keep it up to date without having to go through code reviews from us etc.

**Comment by zoechi** [comment]
Actually I like your review comments :) Ok,  I'll create a repository. If you want to migrate it somewhere else at some point just reach out.

**Comment by jakemac53** [comment]
Sure, and feel free to keep sending me code reviews, I am happy to look at them :)

**Comment by jakemac53** [comment]
closing for now, let me know once these are available elsewhere and I will link on the README, or just send a PR to update the README. Thanks!

**Comment by zoechi** [comment]
Yup, well do. Thanks for the support so far!
